internal/selfupdate/selfupdate.go

```diff
@@ -0,0 +1,338 @@
+	var currentMajor, currentMinor, currentPatch, newMajor, newMinor, newPatch int
+	if _, err := fmt.Sscanf(su.version, "v%d.%d.%d", &currentMajor, &currentMinor, &currentPatch); err != nil {
+		return false, fmt.Errorf("parse current version (%s): %w", su.version, err)
+	}
+	if _, err := fmt.Sscanf(version, "v%d.%d.%d", &newMajor, &newMinor, &newPatch); err != nil {
+		return false, fmt.Errorf("parse new version (%s): %w", version, err)
+	}
+
+	if newMajor > currentMajor {
+		return true, nil
+	}
+	if newMajor == currentMajor && newMinor > currentMinor {
+		return true, nil
+	}
+	if newMajor == currentMajor && newMinor == currentMinor && newPatch > currentPatch {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (su *SelfUpdater) ApplyUpdate(ctx context.Context) error {
+	rel, err := su.checkForUpdates()
+	if err != nil {
+		return err
+	}
+	if rel == nil {
+		return nil
+	}
+
+	if isNewer, err := su.isNewer(rel.Version); err != nil {
+		return err
+	} else if !isNewer {
+		return nil
+	}
+
+	action, err := wailsruntime.MessageDialog(ctx, wailsruntime.MessageDialogOptions{
+		Title:         "Would you like to update Zen?",
+		Message:       rel.Description,
+		Buttons:       []string{"Yes", "No"},
+		Type:          wailsruntime.QuestionDialog,
+		DefaultButton: "Yes",
+		CancelButton:  "No",
+	})
+	if err != nil {
+		log.Printf("error occurred while showing update dialog: %v", err)
+		return err
+	}
+	if action == "No" {
+		log.Printf("aborting update, user declined")
+		return nil
+	}
+
+	ext := filepath.Ext(rel.AssetURL)
+	if strings.HasSuffix(rel.AssetURL, ".tar.gz") {
+		ext = ".tar.gz"
+	}
+
+	if ext != ".tar.gz" && ext != ".zip" {
+		return fmt.Errorf("unsupported archive format: %s", ext)
+	}
+
+	tmpFile, err := os.CreateTemp("", "downloaded-*"+ext)
+	if err != nil {
+		return fmt.Errorf("create temporary file: %v", err)
+	}
+	defer os.Remove(tmpFile.Name())
+
+	err = su.DownloadFile(rel.AssetURL, tmpFile.Name())
+	if err != nil {
+		return fmt.Errorf("download file: %v", err)
+	}
+
+	err = verifyFileHash(tmpFile.Name(), rel.SHA256)
+	if err != nil {
+		return fmt.Errorf("verify file hash: %v", err)
+	}
+
+	var dest string
+	if runtime.GOOS == "darwin" {
+		dest = "/Applications"
+
+		err = removeContents(path.Join(dest, "Zen.app"))
+		if err != nil {
+			return fmt.Errorf("remove contents: %v", err)
+		}
+
+		err = Unarchive(tmpFile.Name(), dest)
+		if err != nil {
+			return fmt.Errorf("unzip file: %v", err)
+		}
+	} else if runtime.GOOS == "windows" {
+		tempUnarchiveDir, err := os.MkdirTemp("", "unarchive-*")
+		if err != nil {
+			return fmt.Errorf("create temp unarchive dir: %v", err)
+		}
+		defer os.RemoveAll(tempUnarchiveDir)
+
+		err = Unarchive(tmpFile.Name(), tempUnarchiveDir)
+		if err != nil {
+			return fmt.Errorf("unzip file: %v", err)
+		}
+
+		var expectedExecName string
+		switch runtime.GOOS {
+		case "windows":
+			expectedExecName = "Zen.exe"
+		case "linux":
```

Comment:
I suspect this will never execute given we're inside the `runtime.GOOS == "windows"` if branch.

Reply:
Rewrote this whole part to switch-case for better readability.